packages/tests/server/methodOverride.test.ts

```diff
@@ -0,0 +1,439 @@
+  batch?: boolean;
+}) {
+  const handler = createHTTPHandler({
+    router: router,
+    unstable_methodOverride: opts.methodOverride,
+  });
+  const requests: {
+    method: string;
+    url: string;
+    body: unknown;
+  }[] = [];
+  const server = http.createServer(async (req, res) => {
+    assert(req.url);
+    assert(req.method);
+    const bodyResult = await getPostBody({ req });
+
+    const body =
+      bodyResult.ok && bodyResult.data !== undefined
+        ? JSON.parse(bodyResult.data as string)
+        : null;
+
+    requests.push({
+      method: req.method,
+      url: req.url,
+      body,
+    });
+
+    (req as any).body = body;
+
+    handler(req, res);
+  });
+
+  server.listen(0);
+
+  const port = (server.address() as any).port as number;
+
+  const client = createTRPCProxyClient<typeof router>({
+    links: [
+      opts.batch
+        ? httpBatchLink({
+            url: `http://localhost:${port}`,
+            fetch: fetch as any,
+            ...opts.linkOptions,
+          })
+        : httpLink({
+            url: `http://localhost:${port}`,
+            fetch: fetch as any,
+            ...opts.linkOptions,
+          }),
+    ],
+  });
+
+  return {
+    port,
+    router,
+    client,
+    requests,
+    close: () => {
+      return new Promise<void>((resolve) => {
+        server.close(() => {
+          resolve();
+        });
+      });
+    },
+  };
+}
+let app: Awaited<ReturnType<typeof startServer>>;
+async function setup(...args: Parameters<typeof startServer>) {
+  app = await startServer(...args);
+  return app;
+}
+
+afterEach(async () => {
+  if (app) {
+    app.close();
+  }
+});
+
+test('normal queries (sanity check)', async () => {
+  const t = await setup({
+    methodOverride: {
+      enabled: false,
+    },
+  });
+
+  expect(
+    await t.client.q.query({
+      who: 'test1',
+    }),
+  ).toBe('hello test1');
+  expect(
+    await t.client.m.mutate({
+      who: 'test2',
+    }),
+  ).toBe('hello test2');
+
+  expect(t.requests.map((req) => req.method)).toEqual(['GET', 'POST']);
+  expect(t.requests).toMatchInlineSnapshot(`
+    Array [
+      Object {
+        "body": null,
+        "method": "GET",
+        "url": "/q?input=%7B%22who%22%3A%22test1%22%7D",
+      },
+      Object {
+        "body": Object {
+          "who": "test2",
+        },
+        "method": "POST",
+        "url": "/m",
+      },
+    ]
+  `);
+});
+
+test('client: sends query as POST when methodOverride=POST', async () => {
+  const t = await setup({
+    methodOverride: {
+      enabled: true,
+    },
+    linkOptions: {
+      unstable_methodOverride: 'POST',
+    },
+  });
+
+  expect(
+    await t.client.q.query({
+      who: 'test1',
+    }),
+  ).toBe('hello test1');
+
+  expect(t.requests).toHaveLength(1);
+  const req = t.requests[0]!;
+
+  expect(req.method).toBe('POST');
+  expect(req.url).toContain('_procedureType=query'); // maybe this should be _type=query? or instead?
```

Comment:
> > I realised that could just skip the override query param too since we can never have collisions on the same key for queries or mutations
> 
> Do you mean we skip the query param and if `methodOverride: true` on the server we just try to find the procedure in both `queries` and `mutations` ?

yeah, I mean could just skip this part :)

Reply:
yep. "downside" is that then you have to check the `def` for both `query` and `mutation` but that's not a huge deal, just a bit less explicit

an alternative API to this whole thing would be to add `httpMethod?: 'GET' | 'POST'` to `ProcedureOptions`, since if we'll try to find the procedure both as a query and mutation on the server anyway, `methodOverride` doesn't need to be a global client-side setting and can be done per procedure call